comparison/comparison.go

```diff
@@ -0,0 +1,509 @@
+	DashboardURL string
+
+	deploymentID string
+}
+
+// New creates and initializes a new Comparison object to be used to run
+// automated load-test comparisons. It returns an error in case of failure.
+func New(cfg *Config, deployerCfg *deployment.Config) (*Comparison, error) {
+	if err := defaults.Validate(cfg); err != nil {
+		return nil, fmt.Errorf("failed to validate config: %w", err)
+	}
+
+	cmp := &Comparison{
+		config:      cfg,
+		deployments: map[string]*deploymentInfo{},
+		cancelCh:    make(chan struct{}),
+	}
+
+	var i int
+	deployments := map[string]string{}
+	for _, lt := range cfg.LoadTests {
+		key := string(lt.Type) + string(lt.DBEngine)
+		var dp *deploymentInfo
+		if id, ok := deployments[key]; !ok {
+			dp = &deploymentInfo{config: *deployerCfg}
+			dp.config.DBInstanceEngine = "aurora-" + string(lt.DBEngine)
+			dp.config.ClusterName = fmt.Sprintf("%s%d", dp.config.ClusterName, i)
+			id = fmt.Sprintf("deployment%d", i)
+			cmp.deployments[id] = dp
+			deployments[key] = id
+			i++
+		} else {
+			dp = cmp.deployments[id]
+		}
+		dp.loadTests = append(dp.loadTests, lt)
+	}
+
+	return cmp, nil
+}
+
+func (c *Comparison) deploymentAction(action func(t *terraform.Terraform) error) error {
+	var wg sync.WaitGroup
+	wg.Add(len(c.deployments))
+	errsCh := make(chan error, len(c.deployments))
+	for id, dp := range c.deployments {
+		go func(id string, dp *deploymentInfo) {
+			defer wg.Done()
+			t := terraform.New(id, &dp.config)
+			defer t.Cleanup()
+			if err := action(t); err != nil {
+				errsCh <- fmt.Errorf("deployment action failed: %w", err)
+			}
+		}(id, dp)
+	}
+	wg.Wait()
+	close(errsCh)
+	var err error
+	for err = range errsCh {
+		mlog.Error(err.Error())
+	}
+	return err
+}
+
+func (c *Comparison) getLoadTestsCount() int {
+	var count int
+	for _, dp := range c.deployments {
+		for range dp.loadTests {
+			count++
+		}
+	}
+	return count
+}
+
+func runBoundedLoadTest(t *terraform.Terraform, coordConfig *coordinator.Config, d time.Duration, cancelCh <-chan struct{}) (coordinator.Status, error) {
+	var err error
+	var status coordinator.Status
+	mlog.Info("starting bounded load-test")
+	if err := t.StartCoordinator(coordConfig); err != nil {
+		return status, err
+	}
+
+	var canceled bool
+	select {
+	case <-cancelCh:
+		mlog.Info("cancelling load-test")
+		canceled = true
+	case <-time.After(d):
+	}
+
+	mlog.Info("stopping bounded load-test")
+	status, err = t.StopCoordinator()
+	if err != nil {
+		return status, err
+	}
+
+	if canceled {
+		return status, errors.New("canceled")
+	}
+
+	// TODO: remove this once MM-30326 has been merged and a new release
+	// published.
+	status.StopTime = time.Now()
+
+	return status, nil
+}
+
+func runUnboundedLoadTest(t *terraform.Terraform, coordConfig *coordinator.Config, cancelCh <-chan struct{}) (coordinator.Status, error) {
+	mlog.Info("starting unbounded load-test")
+	if err := t.StartCoordinator(coordConfig); err != nil {
+		return coordinator.Status{}, err
```

Comment:
Nit: We can probably do `status := coordinator.Status{}` and let that variable get reused everywhere else.

Reply:
Funny enough, it was like that originally and got a nasty bug cause it got shadowed. If we do that we also need to declare an `err` variable to re-use.

Reply:
Lol no worries :)